POST api/usersとGET api/usersの実装

takumi-333 opened the pull request:
fix #66
fix #64 
## 概要
- `POST api/users`の実装
- `GET api/users`の実装

## やったこと
- POSTメソッド
  - setUserの実装
  - setWorkの実装
  - setFieldの実装
  - setAffiliationの実装
- GETメソッド
  - fetchUsersの実装
 
## 特にレビューしてほしいところ
特にないけど、エラーとかミスがないかどうか

## 保留していること
_AfiiliationToUsersおよび_FieldsToUsersの挙動について
それぞれにおいて
全く同じユーザと全く同じ所属先（分野）が登録されたとき、
fetchに失敗する

（そもそも全く同じユーザの登録が二度目起きることはあり得ないが一応報告)

UserDetail型で応急処置
## 動作確認
ローカル環境で動作確認済み

takumi-333 committed:
[add] GET api/users の実装

user情報を一括取得するAPIを実装しました
affiliationIdをクエリとして与えられます
また、GET api/users/[id]同様、UserDetail型を定義して応急処置しています

takumi-333 committed:
[add] POST api/usersの実装

ユーザをAPIを用いて投稿できるようにしました

Miura-code commented on app/api/users/route.ts:
```diff
@@ -0,0 +1,194 @@
+    : null;
+  try {
+    const userDatas = await fetchUsers(affiliationId);
+    if (!userDatas.length) {
```
fetch自体は成功してるから、長さ0のuserDatasをそのまま返していいと思う

takumi-333 replied:
修正しました

Miura-code commented on app/api/users/route.ts:
```diff
@@ -0,0 +1,194 @@
+  try {
+    const newField = await prisma.$queryRaw<Field[]>`
+            INSERT INTO "Fields" (id, name)
+            VALUES (${field.id}, ${field.name})
```
Fieldsテーブルはidが自動的に割り振られるようになっているので明示的に指定する必要はないと思う

Miura-code replied:
→conflictをnameカラムで判断する

takumi-333 replied:
修正しました

Miura-code commented on app/api/users/route.ts:
```diff
@@ -0,0 +1,194 @@
+  try {
+    const newAffiliation = await prisma.$queryRaw<Affiliation[]>`
+            INSERT INTO "Affiliations" (id, name)
+            VALUES (${affiliation.id}, ${affiliation.name})
```
Affiliationテーブルはidが自動的に割り振られるようになっているので明示的にidを指定する必要はないと思う
nameカラムでconflictを判断する

takumi-333 replied:
研究室名が被る可能性を考慮してか、nameカラムのconflictを検知できなかったんよね
データベースの仕様を変える必要がある

Miura-code replied:
なるほど
supabaseの設定見て理解した
個人的には研究室名が被る可能性は、今は考慮する必要はないと思ってる

Miura-code reviewed (commented):
インラインコメントを追加しました

takumi-333 committed:
[fix] 指摘箇所および簡単な修正

setFieldsでは名前のコンフリクト時に既存のものを使うように修正
fetchUserでuserDatas.length == 0でもエラーにしない
works, affiliations, fieldsがnullでもエラーが起きないように修正
